File: SampleSwift/SampleSwift/JoinViewController.swift
//
//  JoinViewController.swift
//  SampleSwift
//
//  Created by Hale Xie on 2020/1/21.
//  Copyright © 2020 Helplightning. All rights reserved.
//

import UIKit
import HLSDK
import HLSDKSwift

let kDefaultUserName = "small_u13";

class JoinViewController: UIViewController {
    @IBOutlet private var gssServerURLTextField: UITextField!
    @IBOutlet private var sessionIDTextField: UITextField!
    @IBOutlet private var sessionPINTextField: UITextField!
    @IBOutlet private var sessionTokenTextView: UITextView!
    @IBOutlet private var userAvatarTextField: UITextField!
    @IBOutlet private var userNameTextField: UITextField!
    @IBOutlet private weak var joinButton: UIButton!
    @IBOutlet private weak var indicator: UIActivityIndicatorView!
    @IBOutlet private weak var themColorPicker: UISegmentedControl!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        sessionPINTextField.text = CallManager.sharedInstance().sessionPIN
        sessionIDTextField.text = CallManager.sharedInstance().sessionID
        sessionTokenTextView.text = CallManager.sharedInstance().sessionToken
        userNameTextField.text = CallManager.sharedInstance().userName;
        userAvatarTextField.text = CallManager.sharedInstance().userAvatar;
        gssServerURLTextField.text = CallManager.sharedInstance().gssServerURL;
        userNameTextField.text = kDefaultUserName;
    }
    
    @IBAction func OnCancel(_ sender: Any) {
        HLClientSwift.shared.stopCurrentCall().always {
            self.presentingViewController?.dismiss(animated: true, completion: nil);
        };
    }
    
    @IBAction func OnJoinCall(_ sender: Any) {
        _setupTheme()
        
        CallManager.sharedInstance().sessionID = sessionIDTextField.text
        CallManager.sharedInstance().sessionToken = sessionTokenTextView.text
        CallManager.sharedInstance().userName = userNameTextField.text
        CallManager.sharedInstance().userAvatar = userAvatarTextField.text
        CallManager.sharedInstance().gssServerURL = gssServerURLTextField.text

        guard let call = HLCall(sessionId: CallManager.sharedInstance().sessionID, sessionToken: CallManager.sharedInstance().sessionToken, userToken: CallManager.sharedInstance().userToken, gssUrl: CallManager.sharedInstance().gssServerURL, localUserDisplayName: CallManager.sharedInstance().userName, localUserAvatarUrl: CallManager.sharedInstance().userAvatar) else { return; }
        
        //  Converted to Swift 5.1 by Swiftify v5.1.33915 - https://objectivec2swift.com/
        joinButton.isEnabled = false
        indicator.isHidden = false
        HLClientSwift.shared.startCall(call: call, viewController: self).then({ value in
            print("The call has started")
        }).catch({ error in
            print("Cannot start the call:\(error)")
        }).always({
            self.indicator.isHidden = true
            self.joinButton.isEnabled = true
        })

    }
    
    func _setupTheme () {
        let index = themColorPicker.selectedSegmentIndex
        var theme: HLTheme? = nil
        switch index {
            case 1:
                theme = HLTheme()
                theme?.setColor(kHLMainColor, color: UIColor.darkGray)
                theme?.setImage(kHLImageModeMenuFaceToFaceOn, image: UIImage(named: "Lightning"))
            case 2:
                theme = HLTheme()
                theme?.setColor(kHLMainColor, color: UIColor.orange)
                theme?.setImage(kHLImageModeMenuFaceToFaceOn, image: UIImage(named: "03_contacts"))
            case 3:
                theme = HLTheme()
                theme?.setColor(kHLMainColor, color: UIColor.purple)
                theme?.setImage(kHLImageModeMenuFaceToFaceOn, image: UIImage(named: "03_contacts_active"))
                theme?.setImage(kHLImageModeMenuFaceToFaceOff, image: UIImage(named: "Lightning"))
            default:
                break
        }
        HLClientSwift.shared.setTheme(theme: theme);
    }

}

File: SampleXamarin/SampleiOS/JoinViewController.cs
﻿using System;

using UIKit;
using System.Threading.Tasks;
using CoreFoundation;
using System.Collections.Generic;

namespace HelpLightning.SDK.Sample.iOS
{
    public partial class JoinViewController : UIViewController, ICallClientDelegate
    {
        private static readonly string DefaultUserName = "small_u13";
        private static readonly string HLApiKey = "[YOUR_HL_API_KEY]";

        public JoinViewController(IntPtr handle) : base(handle)
        {
            CallClientFactory.Instance.CallClient.Delegate = this;
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            sessionPINTextField.Text = CallManager.Instance.SessionPIN;
            sessionIDTextField.Text = CallManager.Instance.SessionID;
            sessionTokenTextView.Text = CallManager.Instance.SessionToken;
            userNameTextField.Text = CallManager.Instance.UserName;
            userAvatarTextField.Text = CallManager.Instance.UserAvatar;
            gssServerURLTextField.Text = CallManager.Instance.GssServerURL;
            userNameTextField.Text = DefaultUserName;

            camOnSwitch.On = CallManager.Instance.AutoEnableCamera;
            micOnSwitch.On = CallManager.Instance.AutoEnableMicrophone;
        }

        public override void DidReceiveMemoryWarning()
        {
            base.DidReceiveMemoryWarning();
            // Release any cached data, images, etc that aren't in use.
        }

        partial void OnJoinCall(UIButton sender)
        {
            SetupTheme();
            CallManager.Instance.SessionID = sessionIDTextField.Text.Trim();
            CallManager.Instance.SessionToken = sessionTokenTextView.Text.Trim();
            CallManager.Instance.UserName = userNameTextField.Text.Trim();
            CallManager.Instance.UserAvatar = userAvatarTextField.Text.Trim();
            CallManager.Instance.GssServerURL = gssServerURLTextField.Text.Trim();
            CallManager.Instance.AutoEnableCamera = camOnSwitch.On;
            CallManager.Instance.AutoEnableMicrophone = micOnSwitch.On;
            Call call = new Call(CallManager.Instance.SessionID,
                CallManager.Instance.SessionToken,
                CallManager.Instance.UserToken,
                CallManager.Instance.GssServerURL,
                HLApiKey,
                CallManager.Instance.UserName,
                CallManager.Instance.UserAvatar,
                CallManager.Instance.AutoEnableCamera,
                CallManager.Instance.AutoEnableMicrophone
                );

            joinButton.Enabled = false;
            indicator.Hidden = false;

            Task<IDictionary<string, object>> task = CallClientFactory.Instance.CallClient.StartCall(call, this);
            task.ContinueWith(t =>
            {
                DispatchQueue.MainQueue.DispatchAsync(() =>
                {
                    if (!t.IsFaulted)
                    {
                        Console.WriteLine("The call has started: " + t.Result[Call.HLCallInfoCallIDKey]);
                    }
                    else
                    {
                        Console.Error.WriteLine("Cannot start the call: " + t.Exception);
                    }

                    joinButton.Enabled = true;
                    indicator.Hidden = true;
                });
            });
        }

        partial void OnCancel(UIBarButtonItem sender)
        {
            this.PresentingViewController.DismissViewController(true, null);
        }

        private void SetupTheme()
        {
            var index = themColorPicker.SelectedSegment;
            Theme theme = null;
            switch (index)
            {
                case 1:
                    theme = new Theme();
                    theme.SetColor(Theme.ColorMain, UIColor.DarkGray);
                    break;

                case 2:
                    theme = new Theme();
                    theme.SetColor(Theme.ColorMain, UIColor.Orange);
                    break;
                case 3:
                    theme = new Theme();
                    theme.SetColor(Theme.ColorMain, UIColor.Purple);
                    theme.SetImage(Theme.ImageMainMenuDocumentOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageMainMenuDocumentOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageMainMenuTorchOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageMainMenuTorchOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageMainMenu, UIImage.FromBundle("Lightning"));

                    theme.SetImage(Theme.ImageModeMenuFaceToFaceOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageModeMenuFaceToFaceOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageModeMenuReceiverOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageModeMenuReceiverOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageModeMenuGiverOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageModeMenuGiverOff, UIImage.FromBundle("Lightning"));

                    theme.SetImage(Theme.ImageCameraMenuPhotoOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuPhotoOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuFreezeOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuFreezeOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuCameraOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuSwitchCamera, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuFrontCameraOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuFrontCameraOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuBackCameraOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuBackCameraOff, UIImage.FromBundle("Lightning"));

                    theme.SetImage(Theme.ImageMicOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageMicOff, UIImage.FromBundle("Lightning"));

                    theme.SetImage(Theme.ImageTelestrationMenuColorUnselected, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageTelestrationMenuColorSelected, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageTelestrationMenuArrowOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageTelestrationMenuArrowOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageTelestrationMenuPenOn, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageTelestrationMenuPenOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageTelestrationUndo, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageTelestrationClearAll, UIImage.FromBundle("Lightning"));

                    theme.SetImage(Theme.ImageEndCall, UIImage.FromBundle("Lightning"));

                    theme.SetImage(Theme.imageScreenCaptureButton1, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.imageScreenCaptureButton2, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.imageScreenCaptureButton3, UIImage.FromBundle("Lightning"));

                    theme.SetImage(Theme.ImageTelestrationMenuArrowSelected, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageTelestrationMenuPenSelected, UIImage.FromBundle("Lightning"));

                    theme.SetImage(Theme.ImageCameraMenuLiveVideoOff, UIImage.FromBundle("Lightning"));
                    theme.SetImage(Theme.ImageCameraMenuLiveVideoOn, UIImage.FromBundle("Lightning"));
                    break;
                default:
                    break;
            }
            CallClientFactory.Instance.CallClient.Theme = theme;
        }

        public void OnCallEnded(Call call, string reason)
        {
            Console.WriteLine("Call Ended");
        }

        public void OnScreenCaptureCreated(Call call, object image)
        {
            Console.WriteLine("Screen Captured");
            imagePreview.Image = (UIImage)image;
        }
    }
}

